src/client.rs

```diff
@@ -280,3 +270,368 @@ where
+                }
+                // Ignore and wait for next status event:
+                _ => continue,
+            }
+        }
+        Err(RpcError::Custom("RPC subscription dropped".into()).into())
+    }
+
+    /// Wait for the transaction to be finalized, and return a [`TransactionInBlock`]
+    /// instance when it is, or an error if there was a problem waiting for finalization.
+    ///
+    /// **Note:** consumes `self`. If you'd like to perform multiple actions as the state of the
+    /// transaction progresses, use [`TransactionProgress::next()`] instead.
+    pub async fn wait_for_finalized(
+        mut self,
+    ) -> Result<TransactionInBlock<'client, T>, Error> {
+        while let Some(status) = self.next().await? {
+            match status {
+                // Finalized! Return.
+                TransactionProgressStatus::Finalized(s) => return Ok(s),
+                // Error scenarios; return the error.
+                TransactionProgressStatus::FinalityTimeout(_) => {
+                    return Err(TransactionError::FinalitySubscriptionTimeout.into())
+                }
+                TransactionProgressStatus::Invalid => {
+                    return Err(TransactionError::Invalid.into())
+                }
+                TransactionProgressStatus::Usurped(_) => {
+                    return Err(TransactionError::Usurped.into())
+                }
+                TransactionProgressStatus::Dropped => {
+                    return Err(TransactionError::Dropped.into())
+                }
+                // Ignore and wait for next status event:
+                _ => continue,
+            }
+        }
+        Err(RpcError::Custom("RPC subscription dropped".into()).into())
+    }
+
+    /// Wait for the transaction to be finalized, and for the transaction events to indicate
+    /// that the transaction was successful. Returns the events associated with the transaction,
+    /// as well as a couple of other details (block hash and extrinsic hash).
+    ///
+    /// **Note:** consumes self. If you'd like to perform multiple actions as progress is made,
+    /// use [`TransactionProgress::next()`] instead.
+    pub async fn wait_for_finalized_success(self) -> Result<TransactionEvents<T>, Error> {
+        let evs = self.wait_for_finalized().await?.wait_for_success().await?;
+        Ok(evs)
+    }
+}
+
+/// Possible transaction progess states returned from our [`TransactionProgress::next()`] call.
+#[derive(Debug)]
+pub enum TransactionProgressStatus<'client, T: Config> {
+    /// Transaction is part of the future queue.
+    Future,
+    /// Transaction is part of the ready queue.
+    Ready,
+    /// The transaction has been broadcast to the given peers.
+    Broadcast(Vec<String>),
+    /// Transaction has been included in block with given hash.
+    InBlock(TransactionInBlock<'client, T>),
+    /// The block this transaction was included in has been retracted.
+    Retracted(T::Hash),
```

Comment:
Is it safe to retry the extrinsic after this? Or do users need to rebuild it from scratch? Either way it'd be good to document.

Reply:
Judging by how the old code worked, if this event happens we expect that the transaction will still make its way into a block eventually.

I'll make a separate issue I think for documenting these things better; the comments all come straight out of substrate, and so I'm a bit hesitant to change them to what I assume they mean without double checking with somebody first (but they are pretty crap, and could def be improved)!

Reply:
So I dug into the substrate docs and I found a useful comment on the enum itself which I've adapted and will put into this, which I think helps to clear up some of the confusion (but also makes it more ambiguous in some ways...)